File: python/agentbay/_common/params/lifecycle_policy.py
from typing import Optional


class LifecyclePolicy:
    """
    Lifecycle policy for session management.

    Controls how and when a session is automatically released.
    When used, SDK takes full control of lifecycle — console defaults are overridden.

    All time values are in MINUTES.

    Attributes:
        idle_release_timeout: Minutes of inactivity before auto-release (default: 5).
        max_runtime: Maximum total runtime in minutes from creation (default: 30).
        manual_release: If True, disables all auto-release; session only ends via delete().
    """

    def __init__(
        self,
        idle_release_timeout: Optional[int] = None,
        max_runtime: Optional[int] = None,
        manual_release: bool = False,
    ):
        if manual_release:
            if idle_release_timeout is not None:
                raise ValueError(
                    "idle_release_timeout cannot be set when manual_release=True. "
                    "In manual release mode, the session is only released via delete()."
                )
            if max_runtime is not None:
                raise ValueError(
                    "max_runtime cannot be set when manual_release=True. "
                    "In manual release mode, the session is only released via delete()."
                )
            self.idle_release_timeout = 0
            self.max_runtime = 0
            self.manual_release = True
            return

        resolved_idle = idle_release_timeout if idle_release_timeout is not None else 5
        resolved_max = max_runtime if max_runtime is not None else 30

        if not isinstance(resolved_idle, int) or isinstance(resolved_idle, bool):
            raise ValueError("idle_release_timeout must be a positive integer (minutes)")
        if resolved_idle <= 0:
            raise ValueError("idle_release_timeout must be a positive integer (minutes)")

        if not isinstance(resolved_max, int) or isinstance(resolved_max, bool):
            raise ValueError("max_runtime must be a positive integer (minutes)")
        if resolved_max <= 0:
            raise ValueError("max_runtime must be a positive integer (minutes)")

        self.idle_release_timeout = resolved_idle
        self.max_runtime = resolved_max
        self.manual_release = False

File: typescript/src/lifecycle-policy.ts
/**
 * Options for creating a LifecyclePolicy.
 */
export interface LifecyclePolicyOptions {
  /** Idle release timeout in minutes (default: 5). Must be a positive integer. */
  idleReleaseTimeout?: number;
  /** Maximum session runtime in minutes from creation (default: 30). Must be a positive integer. */
  maxRuntime?: number;
  /** When true, disables all auto-release; the session only ends via delete(). */
  manualRelease?: boolean;
}

/**
 * Lifecycle policy for session management.
 *
 * Controls how and when a session is automatically released.
 * When used, SDK takes full control of lifecycle — console defaults are overridden.
 * All time values are in MINUTES.
 *
 * Three control dimensions:
 * - **idleReleaseTimeout**: Minutes of inactivity before auto-release (default: 5)
 * - **maxRuntime**: Absolute maximum session duration from creation (default: 30)
 * - **manualRelease**: Disable all auto-release; session only ends via `delete()`
 */
export class LifecyclePolicy {
  /** Minutes of inactivity before auto-release (default: 5). */
  readonly idleReleaseTimeout: number;
  /** Maximum session runtime in minutes from creation (default: 30). */
  readonly maxRuntime: number;
  /** When true, disables all auto-release; session only ends via delete(). */
  readonly manualRelease: boolean;

  constructor(options?: LifecyclePolicyOptions) {
    const opts = options || {};

    if (opts.manualRelease) {
      if (
        opts.idleReleaseTimeout !== undefined &&
        opts.idleReleaseTimeout !== null
      ) {
        throw new Error(
          "idleReleaseTimeout cannot be set when manualRelease=true. " +
            "In manual release mode, the session is only released via delete()."
        );
      }
      if (opts.maxRuntime !== undefined && opts.maxRuntime !== null) {
        throw new Error(
          "maxRuntime cannot be set when manualRelease=true. " +
            "In manual release mode, the session is only released via delete()."
        );
      }
      this.idleReleaseTimeout = 0;
      this.maxRuntime = 0;
      this.manualRelease = true;
      return;
    }

    const idle =
      opts.idleReleaseTimeout !== undefined && opts.idleReleaseTimeout !== null
        ? opts.idleReleaseTimeout
        : 5;
    const max =
      opts.maxRuntime !== undefined && opts.maxRuntime !== null
        ? opts.maxRuntime
        : 30;

    if (typeof idle !== "number" || !Number.isInteger(idle) || idle <= 0) {
      throw new Error(
        "idleReleaseTimeout must be a positive integer (minutes)"
      );
    }
    if (typeof max !== "number" || !Number.isInteger(max) || max <= 0) {
      throw new Error("maxRuntime must be a positive integer (minutes)");
    }

    this.idleReleaseTimeout = idle;
    this.maxRuntime = max;
    this.manualRelease = false;
  }
}
